### PR TITLE
[learning-rate-scheduler] New interface for learning rate scheduler 

### PR DESCRIPTION
**Pull request by kparichay**
- Move declaration and definition of learning rate related properties out
to common properties from optimizer_impl class.
- Add interface for the learning rate scheduler which all learning rate
schedulers must abide by.
- Support constant learning rate scheduler.
- Support exponential decay learning rate scheduler with decay rate and
decay steps are the controllable properties.

See Also #1776

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: nntrainer/optimizers/lr_scheduler.h [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+   * @brief     get Learning Rate for the given iteration
+   * @param[in] iteration Iteration for the learning rate
+   * @retval    Learning rate in double
+   * @detail    the return value of this function and getInitialLearningRate()
+   * may not match for iteration == 0 (warmup can lead to different initial
+   * learning rates).
+   *
+   * @note this is non-const function intentionally.
+   */
+  virtual double getLearningRate(size_t iteration) = 0;
```
Comment (kparichay) [inline nntrainer/optimizers/lr_scheduler.h:46]: the current interface does not have `finalize()` to check if the required properties are provided, etc and they throw in the first call to `getLearningRate()`. If you guys think `finalize()` would be better, I will add with interface `void finalize();` and it will throw if any error.

File: nntrainer/optimizers/lr_scheduler.h [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ *
+ */
+
+#ifndef __LEARNING_RATE_SCHEDULER__
+#define __LEARNING_RATE_SCHEDULER__
+#ifdef __cplusplus
+
+#include <string>
+
+#include <node_exporter.h>
```
Comment (zhoonit) [inline nntrainer/optimizers/lr_scheduler.h:20]: Could you put this inside cpp? this is *.hpp and quite heavy.

we will need to fwd declare like below

https://github.com/nnstreamer/nntrainer/blob/e333b7ffe402eb3572b3082f0bfc8b103ba58a4e/nntrainer/optimizers/optimizer_devel.h#L27-L28

File: nntrainer/optimizers/lr_scheduler.h (hunk 1 above)
Comment (zhoonit) [reply nntrainer/optimizers/lr_scheduler.h:46]: +1 for this, this might cause problem some later :p

**Review by zhoonit — APPROVED** [review]
LGTM :)

Does not need to do within this PR but we will need a simple test to verify the behavior.

File: nntrainer/optimizers/lr_scheduler.h (hunk 2 above)
Comment (kparichay) [reply nntrainer/optimizers/lr_scheduler.h]: done.

File: nntrainer/optimizers/lr_scheduler.h (hunk 1 above)
Comment (kparichay) [reply nntrainer/optimizers/lr_scheduler.h:57]: added finalize with popular demand :)

**Comment by kparichay** [comment]
V2: Add finalize to the interface of the learning rate scheduler with the purpose to verify that the required properties have been set.

**Review by zhoonit — APPROVED** [review]
looks good :)

File: nntrainer/optimizers/lr_scheduler_exponential.h [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+#ifdef __cplusplus
+
+#include <string>
+
+#include <lr_scheduler_constant.h>
+
+namespace nntrainer {
+
+/**
+ * @class   Constant Learning Rate Scheduler class
```
Comment (lhs8928) [inline nntrainer/optimizers/lr_scheduler_exponential.h:25]: Constant => exponential

File: nntrainer/optimizers/lr_scheduler_exponential.h [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+ * @brief   class for constant Learning Rate Schedulers
+ */
+class ExponentialLearningRateScheduler : public ConstantLearningRateScheduler {
+
+public:
+  /**
+   * @brief Construct a new constant learning rate scheduler object
```
Comment (lhs8928) [inline nntrainer/optimizers/lr_scheduler_exponential.h:33]: same here

File: nntrainer/layers/common_properties.h [hunk 5]
```diff
@@ -971,6 +971,37 @@ class ClipGradByGlobalNorm : public Property<float> {
...
+
+/**
+ * @brief decay steps property
+ *
+ */
+class DecaySteps : public PositiveIntegerProperty {
+public:
+  static constexpr const char *key = "decay_steps"; /**< unique key to access */
+  using prop_tag = uint_prop_tag;                   /**< property type */
+};
```
Comment (lhs8928) [inline nntrainer/layers/common_properties.h:1003]: I think valid function is needed to check whether the decay_step is not 0

**Review by lhs8928 — APPROVED** [review]
Other than minor comment, looks good to me :)

File: nntrainer/optimizers/lr_scheduler_exponential.h (hunk 4 above)
Comment (kparichay) [reply nntrainer/optimizers/lr_scheduler_exponential.h:33]: Added in later PR, so I will skip here for faster merge.

File: nntrainer/optimizers/lr_scheduler_exponential.h (hunk 3 above)
Comment (kparichay) [reply nntrainer/optimizers/lr_scheduler_exponential.h:25]: Added in later PR, so I will skip here for faster merge.

File: nntrainer/optimizers/optimizer_impl.cpp [hunk 6]
```diff
@@ -12,10 +12,11 @@
  *
  */
 
+#include <cmath>
 #include <fstream>
 #include <iostream>
 
-#include <cmath>
+#include <common_properties.h>
```
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/optimizers/optimizer_impl.cpp:19]: `optimizer_impl.cpp` file is eventually removed in later commits, so I think this can be left like this?

**Review by jijoongmoon — APPROVED** [review]
LGTM
